actors/builtin/miner/miner_test.go

```diff
@@ -390,6 +390,97 @@ func TestProvingPeriodCron(t *testing.T) {
 	})
 }
 
+func TestExtendSectorExpiration(t *testing.T) {
+	owner := tutil.NewIDAddr(t, 100)
+	worker := tutil.NewIDAddr(t, 101)
+	workerKey := tutil.NewBLSAddr(t, 0)
+	actor := newHarness(t, tutil.NewIDAddr(t, 1000), owner, worker, workerKey)
+	periodOffset := abi.ChainEpoch(100)
+	builder := mock.NewBuilder(context.Background(), actor.receiver).
+		WithActorType(owner, builtin.AccountActorCodeID).
+		WithActorType(worker, builtin.AccountActorCodeID).
+		WithHasher(fixedHasher(uint64(periodOffset))).
+		WithCaller(builtin.InitActorAddr, builtin.InitActorCodeID)
+
+	t.Run("rejects negative extension", func(t *testing.T) {
+		rt := builder.Build(t)
+		actor.constructAndVerify(rt, periodOffset)
+		precommitEpoch := abi.ChainEpoch(1)
+		rt.SetEpoch(precommitEpoch)
+		st := getState(rt)
+		deadline := st.DeadlineInfo(rt.Epoch())
+		expiration := deadline.PeriodEnd() + 10*miner.WPoStProvingPeriod
+		sectorInfo := actor.commitAndProveSectors(rt, 1, expiration, big.Zero())
+
+		sector, found, err := getState(rt).GetSector(rt.AdtStore(), sectorInfo[0].SectorNumber)
+		require.NoError(t, err)
+		require.True(t, found)
+
+		// attempt to shorten epoch
+		newExpiration := sector.Info.Expiration - abi.ChainEpoch(42)
+		params := &miner.ExtendSectorExpirationParams{
+			SectorNumber:  sector.Info.SectorNumber,
+			NewExpiration: newExpiration,
+		}
+
+		func() {
+			defer func() {
+				r := recover()
+				require.NotNil(t, r)
+				assert.Contains(t, fmt.Sprintf("%s",r), "cannot reduce sector expiration")
+			}()
+
+			actor.extendSector(rt, sector, 0, params)
+		}()
+
+	})
+
+	t.Run("updates expiration with valid params", func(t *testing.T) {
```

Comment:
Since the set-up for these tests is lengthy, I think it's prudent to combine multiple variations into a single run, where the implicit state dependencies are minimal. That also makes the essential difference between the cases easier to spot. Since aborted calls don't change state at all, you can combine as many of them as you want before a state-changing call.

In this case I suggest combining with the case above. See TestCommitments as example.

Reply:
I have a PR stacked on this one. I'll do this refactor in the followup.